contracts/carbon/Strategies.sol

```diff
@@ -390,118 +398,130 @@ abstract contract Strategies is Initializable {
 
             (Order memory targetOrder, Order memory sourceOrder) = isTargetToken0 == ordersInverted
                 ? (orders[1], orders[0])
                 : (orders[0], orders[1]);
 
             // calculate the orders new values
-            SourceAndTargetAmounts memory tempTradeAmounts = _singleTradeActionSourceAndTargetAmounts(
+            (uint128 sourceAmount, uint128 targetAmount) = _singleTradeActionSourceAndTargetAmounts(
                 targetOrder,
-                tradeActions[i].amount,
+                amount,
                 params.byTargetAmount
             );
 
             // handled specifically for a custom error message
-            if (targetOrder.y < tempTradeAmounts.targetAmount) {
+            if (targetOrder.y < targetAmount) {
                 revert InsufficientLiquidity();
             }
 
             // update the orders with the new values
             // safe since it's checked above
             unchecked {
-                targetOrder.y -= tempTradeAmounts.targetAmount;
+                targetOrder.y -= targetAmount;
             }
 
-            sourceOrder.y += tempTradeAmounts.sourceAmount;
+            sourceOrder.y += sourceAmount;
             if (sourceOrder.z < sourceOrder.y) {
                 sourceOrder.z = sourceOrder.y;
             }
 
             // store new values if necessary
             uint256[3] memory newPackedOrders = _packOrders(orders, ordersInverted);
-            for (uint256 n = 0; n < 3; n = uncheckedInc(n)) {
-                if (packedOrdersMemory[n] != newPackedOrders[n]) {
-                    packedOrders[n] = newPackedOrders[n];
-                }
+
+            // both y values are in slot 0, so it has definitely changed
+            packedOrders[0] = newPackedOrders[0];
+
+            // one of the z values is in slot 1, so it has possibly changed
+            if (packedOrdersMemory[1] != newPackedOrders[1]) {
+                packedOrders[1] = newPackedOrders[1];
+            }
+
+            // the other z value has possibly changed only if the first one hasn't
+            else if (packedOrdersMemory[2] != newPackedOrders[2]) {
+                packedOrders[2] = newPackedOrders[2];
             }
 
             // emit update events if necessary
-            Token[2] memory sortedTokens = _sortStrategyTokens(params.pair, ordersInverted);
-            emit StrategyUpdated({
-                id: strategyId,
-                token0: sortedTokens[0],
-                token1: sortedTokens[1],
-                order0: orders[0],
-                order1: orders[1],
-                reason: STRATEGY_UPDATE_REASON_TRADE
-            });
-
-            totals.sourceAmount += tempTradeAmounts.sourceAmount;
-            totals.targetAmount += tempTradeAmounts.targetAmount;
+            if (ordersInverted) {
+                emit StrategyUpdated({
+                    id: strategyId,
+                    token0: params.pair.tokens[1],
```

Comment:
You should consider using `ordersInverted` inline (i.e., `token0: params.pair.tokens[ordersInverted ? 1 : 0]`) to reduce the impact on the size of the contract.

Reply:
That would yield 2 branches (for token0 and for token1) **per order** (since it's inside the 'for' loop).
If I remember correctly, it doesn't impact the contract size significantly.

Reply:
AFAIK, the "duplication" of the whole `emit` is worse, but we should measure and compare in any case.

Reply:
I'll do that

Reply:
23.133 with two "emits".
23.100 with one "emit".

Reply:
same gas cost, so I'd keep a single branch for readability

Reply:
This specific change even increases the gas cost (very slightly).
It is actually in favor of reducing stack size, in order to allow for the other (gas-related) changes to be compilable.
Otherwise, you get stack-too-deep, which I couldn't find any way to resolve other than with `viaIR: true`.
Can probably resolve by moving into a separate function, if you wish.

Reply:
Cannot resolve with a separate function.

Reply:
Any other alternatives?

Reply:
What are you trying to do? You don't have to use new variables for that, but can use the ternary `? :` instead, no?

Reply:
Option 1:
Get rid of most of the other optimizations in this function, specifically where more local variables are used, for example, getting rid of `uint128 amount = tradeActions[i].amount` AND reverting `_singleTradeActionSourceAndTargetAmounts` to return a memory struct instead of two variables. This would revert a  quite-significant gas reduction.

Option 2:
Check the condition twice instead of once:
```
emit StrategyUpdated({
    id: strategyId,
    token0: params.pair.tokens[ordersInverted ? 1 : 0],
    token1: params.pair.tokens[ordersInverted ? 0 : 1],
    order0: orders[0],
    order1: orders[1],
    reason: STRATEGY_UPDATE_REASON_TRADE
});
```
Which increases the gas cost by either 92 or 140 (depending on the type of trade) gas per order.

Option 3:
Change the event prototype to take one `Token[2]` parameter instead of two `Token` parameters.
This is an API-breaking change of course.

Option 4:
Reduce the event even further, and remove these two tokens altogether, since they can be retrieved also from the strategy ID itself (which contains the pool ID).
This is also an API-breaking change of course.

Reply:
> What are you trying to do? You don't have to use new variables for that, but can use the ternary `? :` instead, no?

Correct. This would yield two checks of `ordersInverted` instead of one (which is the current implementation on mainnet). See the impact of this change in the comment above.

Reply:
Proceeded with option 2.
Please review.